3115-thumbnails-parameters

viktor-wolf opened the pull request:
The thumbnail quality setting works now. I have also added a `$quality` parameter as the last parameter of the `thumbnail` filter. I have also brought the quality setting in `GeneralParser`'s default config in line with the default in `config.yaml`. As a result of my changes, quality is now the 3rd element of generated thumbnail paths. 

I have removed superfluous settings from `config.yaml` (`cropping` and `allow_upscale` – both seem to be superseded by the `$fit` parameter of the `thumbnail` filter).

I have tried to make my changes as undisruptive as possible, but I still recommend giving them a good look over.

bobdenotter reviewed (commented):
Hi @viktor-wolf. Thanks or the PR. 

Why did you choose to make quality the third param? Doesn't that break existing thumbs that have a `fit`? like: 

```
https://playground.wip/thumbs/600×361×f/stock/image_19493.jpg
https://playground.wip/thumbs/600×361×s/stock/image_19493.jpg
```

viktor-wolf commented:
Hi,

my thought process went like this:

1. I thought it would be nice to be able to pass quality as an argument to the `thumbnail` function.
2. I have added `$quality` as the last parameter of the function, in order to preserve backwards compatibility.
3. In `ThumbnailHelper` I have added it as the third argument when `paramString` is being constructed, because it seemed like the most straightforward way of passing it to `ImageController` without needing to completely overhaul the parsing of `fit` and `location` there.
4. I have incremented the indices that are being checked/used as values while parsing the `paramString` by 1. Since quality is always included in the `paramString` (by setting it in a template, or setting it in the config, or from `GeneralParser` default settings), this should always work and preserve backwards compatibility.

Since then, we have talked on Slack, and you said the quality parameter on the `thumbnail` function probably shouldn't be included. If we decide to remove it, we can remove all the changes to the function and `ThumbnailHelper`, and simply pull the quality from configs directly in `ImageController`.

bobdenotter reviewed (approved):
Ok, cool! Let's merge! 👍

viktor-wolf commented:
Hi, thanks for approving the merge request. However, there a couple of loose ends still remain:

1. I was waiting for you to confirm that I should remove the thumbnail parameter from the thumbnail filter/function. Have you decided that it should be left in?
2. Are there any post-merge tasks I should take care of? Updating the docs and such?
